### ISSUE TITLE
Wrong error indication when a type has multiple implementations of a trait via generics.

### ISSUE THREAD

**Issue by Jason5Lee** [intellij-rust/intellij-rust#3295]
## Environment

* **Intellij-Rust plugin version:** v0.2.0.2114-183
* **Rust toolchain version:** rustup 1.16.0 (beab5ac2b 2018-12-06)
* **IDE name and version:** IntelliJ IDEA Community 2018.3.3
* **Operating system:** macOS Mojave

## Problem description

The problem first appeared when I worked with `BorrowMut` trait. A simplified version of the code was

```rust
use std::borrow::BorrowMut;

trait Set<T> {
    fn set(&mut self, new: T);
}

impl<T, B: BorrowMut<T>> Set<T> for B {
    fn set(&mut self, new: T) {
        *self.borrow_mut() = new
    }
}

fn main() {
    let mut a = Box::new(12);
    a.set(24);
    a.set(Box::new(14)); // ERROR
}
```

Then the editor gave an error indication.
![2019-01-26 4 45 27](https://user-images.githubusercontent.com/25118271/51784828-d89e4d00-2189-11e9-8b7a-2edb15663b55.png)
But this code was compiled and just worked fine.

Then I try to reproduced this error without `BorrowMut` trait, and I got this.

```rust
trait Foo<T> {
    fn foo(&self, arg: T);
}

struct Bar<T>(T);

impl<T> Foo<T> for T {
    fn foo(&self, _arg: T) {
        println!("T")
    }
}

impl<T> Foo<T> for Bar<T> {
    fn foo(&self, _arg: T) {
        println!("Bar<T>")
    }
}
fn main() {
    let a = Bar(0);
    a.foo(0);
    a.foo(Bar(0)); // ERROR
}
```

The same problem appeared. The editor complained that the last statement has an error,
![2019-01-26 5 03 11](https://user-images.githubusercontent.com/25118271/51785010-4cd9f000-218c-11e9-87de-46eb4dc49f2a.png)
while the code was compiled correctly and output
```
Bar<T>
T
```
just as the expectation.

I guess when a trait is implemented for a generic type, and one type has multiple implementations of the trait via different generic type, the plugin cannot infer the implementation correctly.
